### DISCUSSION TITLE
Cannot return null for non-nullable field Subscription.greeting

### DISCUSSION THREAD

**Discussion by jmc420** [enisdenjo/graphql-ws#460] [Q&A] [answered]
I am getting "Cannot return null for non-nullable field Subscription.greeting" returned to Postman when I invoke a subscription:

Schema looks like this:

```
const gql = `
    type Query {
        sayHello: String!
      }
      
      type Mutation {
        sendMessage(message:String!):String !
      }
      
      type Subscription {
          greeting: String!
      }
    `;
```
The resolver looks like this:

```
const rootValue = {
    subscription: {
        greeting: async function* () {
            for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
                yield { greetings: hi };
            }
        }
    }
}
```
The full code of the example can be found here:

https://github.com/jmc420/graphql_examples/blob/main/graphql-ws/src/server.ts

**Comment by enisdenjo** [comment]
Try changing your root resolver to use the `subscribe` field instead:

```diff
const rootValue = {
    subscription: {
-        greeting: async function* () {
-            for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
-                yield { greetings: hi };
-            }
-        }
+        greeting: {
+            subscribe: async function* () {
+                for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
+                    yield { greetings: hi };
+                }
+            }
+        }
    }
}
```

**Reply by jmc420** [reply]
I changed it to this and I now get this error in the Node console:

 Error: Subscription field must return Async Iterable. Received: { subscribe: [function subscribe] }.

**Reply by enisdenjo** [reply]
Can you share the code? Could be that you're missing the `*` in the function definition `subscribe: async function* ()`.

**Reply by jmc420** [reply]
The code repo is here:

https://github.com/jmc420/graphql_examples

the code is in the sub directory graphql-ws

The function is marked with *.

**Reply by enisdenjo** [reply]
Ah, its was because you're yielding the field "greeting**s**" while the field is called just "greeting" (singular). Change your original code:

```diff
const rootValue = {
    subscription: {
        greeting: async function* () {
            for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
-                yield { greetings: hi };
+                yield { greeting: hi };
            }
        }
    }
}
```

**Reply by jmc420** [reply]
I changed that and checked it in but am still getting the error.

**Reply by enisdenjo** [reply]
> Change your **original** code

I meant to also remove the `subscribe` field:

```diff
const rootValue = {
    subscription: {
-        greeting: {
-            subscribe: async function* () {
-                for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
-                    yield { greetings: hi };
-                }
-            }
-        }
+        greeting: async function* () {
+            for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
+                yield { greeting: hi };
+            }
+        }
    }
}
```

**Reply by jmc420** [reply]
yes removing the subscribe fixes it. The resolvers for graphql-http don't work in the same way as the resolvers for graphql-ws.

The resolvers for graphql-ws are:

```
const rootValue = {
    query: {
        sayHello: () => {
            return "Hello"
        }
    },
    mutation: {
        sendMessage: ({ message }) => {
            return "You said " + message
        }
    },
    subscription: {
        greeting: async function* () {
            for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
                yield { greeting: hi };
            }
        }
    }
}
```
The resolvers for graphql-http are:

```
const resolvers = {
    sayHello: () => {
        return "Hello"
    },
    sendMessage: ({ message }) => {
        return "You said " + message
    }
}

```

it's rather confusing!

**Reply by enisdenjo** [reply]
You should always use the GraphQL schema and have all resolvers contained within - don't use the rootValue. I heavily recommend `graphql-tools` for assembling the schema. I created a PR for you: https://github.com/jmc420/graphql_examples/pull/1.

**Reply by jmc420** [reply]
Thanks for the pull request. Graphql tools makes things a lot clearer. 

Curiously though the query and the subscription now work, there is a problem with the mutation

```
mutation SendMessage {
    sendMessage(message: "Hello again")
}
```

now generates the error:

"message": "Cannot destructure property 'message' of 'undefined' as it is undefined.",

I added a debug-server command to scripts in package.json. When I put a break point in the body of sendMessage, the breakpoint was not triggered when there was a destructuring parameter; it never reached the code because of the destructuring problem. If I removed the destructuring parameter, the breakpoint was triggered. If I add a parameter to the query, I get the same problem.

Do you think the destructuring problem is within graphql-tools or in graphql-http?

**Reply by enisdenjo** [reply]
> Do you think the destructuring problem is within graphql-tools or in graphql-http?

Neither, the problem is in your code [here](https://github.com/jmc420/graphql_examples/blob/7c69408cfb0d3dae7fcd7df3a7c6cde211d1d8aa/graphql-ws/src/server.ts#L33). The field args are the second argument of the resolver. [Please read the graphql-tools documentation about schema generation.](https://the-guild.dev/graphql/tools/docs/generate-schema)

```diff
Mutation: {
-    sendMessage: ({ message }) => {
+    sendMessage: (_, { message }) => {
        return "You said " + message
    }
},
```

**Reply by jmc420** [reply]
thanks that fixes it.

**Comment by enisdenjo** [comment, marked as answer]
_From https://github.com/enisdenjo/graphql-ws/discussions/460#discussioncomment-5234026_

Its was because of yielding the field "greeting**s**" while the actual field is called just "greeting" (singular).

```diff
const rootValue = {
    subscription: {
        greeting: async function* () {
            for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
-                yield { greetings: hi };
+                yield { greeting: hi };
            }
        }
    }
}
```
